File: sdks/java/src/main/java/com/bcs/BcsDeserializer.java
package com.bcs;

import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.CharsetDecoder;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

/**
 * BCS Deserializer - Manual deserialization API.
 *
 * <p>Provides explicit methods for deserializing each BCS type.
 * Use this for full control over deserialization.</p>
 *
 * <pre>{@code
 * byte[] data = ...;
 * BcsDeserializer des = new BcsDeserializer(data);
 * short value1 = des.readU8();
 * long value2 = des.readU64();
 * String value3 = des.readString();
 * des.checkEnd();
 * }</pre>
 *
 * @implNote Uses a ThreadLocal CharsetDecoder for efficient string validation.
 */
public class BcsDeserializer {

    // Pre-computed constants for signed integer conversion
    private static final BigInteger SIGN_BIT_128 = BigInteger.ONE.shiftLeft(127);
    private static final BigInteger TWO_POW_128 = BigInteger.ONE.shiftLeft(128);
    private static final BigInteger SIGN_BIT_256 = BigInteger.ONE.shiftLeft(255);
    private static final BigInteger TWO_POW_256 = BigInteger.ONE.shiftLeft(256);

    // ThreadLocal CharsetDecoder for efficient string validation (avoids allocation per call)
    private static final ThreadLocal<CharsetDecoder> UTF8_DECODER = ThreadLocal.withInitial(() ->
        StandardCharsets.UTF_8.newDecoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT)
    );

    private final byte[] data;
    private int offset = 0;
    private int currentDepth = 0;

    /**
     * Create a new deserializer.
     *
     * @param data the data to deserialize from
     */
    public BcsDeserializer(byte[] data) {
        this.data = data;
    }

    // ==========================================================================
    // BOOLEAN
    // ==========================================================================

    /**
     * Deserialize a boolean value.
     *
     * @return the boolean value
     * @throws BcsError if invalid
     */
    public boolean readBool() {
        ensureBytes(1);
        int byteVal = data[offset++] & 0xFF;
        if (byteVal == 0) {
            return false;
        }
        if (byteVal == 1) {
            return true;
        }
        throw BcsError.invalidBoolean(byteVal);
    }

    // ==========================================================================
    // UNSIGNED INTEGERS
    // ==========================================================================

    /**
     * Deserialize an unsigned 8-bit integer.
     *
     * @return the value (0-255)
     */
    public short readU8() {
        ensureBytes(1);
        return (short) (data[offset++] & 0xFF);
    }

    /**
     * Deserialize an unsigned 16-bit integer (little-endian).
     *
     * @return the value (0-65535)
     */
    public int readU16() {
        ensureBytes(2);
        int value = (data[offset] & 0xFF) | ((data[offset + 1] & 0xFF) << 8);
        offset += 2;
        return value;
    }

    /**
     * Deserialize an unsigned 32-bit integer (little-endian).
     *
     * @return the value (0 to 2^32-1)
     */
    public long readU32() {
        ensureBytes(4);
        long value = (data[offset] & 0xFFL)
                | ((data[offset + 1] & 0xFFL) << 8)
                | ((data[offset + 2] & 0xFFL) << 16)
                | ((data[offset + 3] & 0xFFL) << 24);
        offset += 4;
        return value;
    }

    /**
     * Deserialize an unsigned 64-bit integer (little-endian).
     *
     * @return the value (as long, interpret as unsigned)
     */
    public long readU64() {
        ensureBytes(8);
        // Inline for performance - avoid method call overhead
        long value = (data[offset] & 0xFFL)
                | ((data[offset + 1] & 0xFFL) << 8)
                | ((data[offset + 2] & 0xFFL) << 16)
                | ((data[offset + 3] & 0xFFL) << 24)
                | ((data[offset + 4] & 0xFFL) << 32)
                | ((data[offset + 5] & 0xFFL) << 40)
                | ((data[offset + 6] & 0xFFL) << 48)
                | ((data[offset + 7] & 0xFFL) << 56);
        offset += 8;
        return value;
    }

    /**
     * Deserialize an unsigned 128-bit integer (little-endian).
     *
     * @return the value as BigInteger
     */
    public BigInteger readU128() {
        ensureBytes(16);
        return readBigIntegerLE(16);
    }

    /**
     * Deserialize an unsigned 256-bit integer (little-endian).
     *
     * @return the value as BigInteger
     */
    public BigInteger readU256() {
        ensureBytes(32);
        return readBigIntegerLE(32);
    }

    // ==========================================================================
    // SIGNED INTEGERS
    // ==========================================================================

    /**
     * Deserialize a signed 8-bit integer (two's complement).
     *
     * @return the value (-128 to 127)
     */
    public byte readI8() {
        ensureBytes(1);
        return data[offset++];
    }

    /**
     * Deserialize a signed 16-bit integer (two's complement, little-endian).
     *
     * @return the value (-32768 to 32767)
     */
    public short readI16() {
        int unsigned = readU16();
        return (short) unsigned;
    }

    /**
     * Deserialize a signed 32-bit integer (two's complement, little-endian).
     *
     * @return the value
     */
    public int readI32() {
        long unsigned = readU32();
        return (int) unsigned;
    }

    /**
     * Deserialize a signed 64-bit integer (two's complement, little-endian).
     *
     * @return the value
     */
    public long readI64() {
        return readU64(); // Same bits, Java interprets as signed
    }

    /**
     * Deserialize a signed 128-bit integer (two's complement, little-endian).
     *
     * @return the value as BigInteger
     */
    public BigInteger readI128() {
        BigInteger unsigned = readU128();
        if (unsigned.compareTo(SIGN_BIT_128) >= 0) {
            return unsigned.subtract(TWO_POW_128);
        }
        return unsigned;
    }

    /**
     * Deserialize a signed 256-bit integer (two's complement, little-endian).
     *
     * @return the value as BigInteger
     */
    public BigInteger readI256() {
        BigInteger unsigned = readU256();
        if (unsigned.compareTo(SIGN_BIT_256) >= 0) {
            return unsigned.subtract(TWO_POW_256);
        }
        return unsigned;
    }

    // ==========================================================================
    // ULEB128
    // ==========================================================================

    /**
     * Deserialize a ULEB128-encoded unsigned integer.
     *
     * @return the value (0 to 2^32-1)
     */
    public long readUleb128() {
        Uleb128.DecodeResult result = Uleb128.decode(data, offset);
        offset += result.getBytesConsumed();
        return result.getValue();
    }

    // ==========================================================================
    // BYTES AND STRINGS
    // ==========================================================================

    /**
     * Deserialize a byte array (length-prefixed with ULEB128).
     *
     * @return the bytes
     */
    public byte[] readBytes() {
        long length = readUleb128();
        if (length > BcsSerializer.MAX_SEQUENCE_LENGTH) {
            throw BcsError.exceededMaxLength(length);
        }
        // Defense-in-depth: explicit check before cast
        if (length > Integer.MAX_VALUE) {
            throw BcsError.exceededMaxLength(length);
        }
        return readFixedBytes((int) length);
    }

    /**
     * Deserialize a UTF-8 string (length-prefixed with ULEB128).
     *
     * @return the string
     */
    public String readString() {
        byte[] bytes = readBytes();
        // Use cached ThreadLocal decoder for better performance
        CharsetDecoder decoder = UTF8_DECODER.get();
        decoder.reset(); // Reset state for reuse
        try {
            return decoder.decode(ByteBuffer.wrap(bytes)).toString();
        } catch (CharacterCodingException e) {
            throw BcsError.invalidUtf8(e.getMessage());
        }
    }

    /**
     * Deserialize fixed-length bytes (no length prefix).
     *
     * @param length the number of bytes to read
     * @return the bytes
     */
    public byte[] readFixedBytes(int length) {
        ensureBytes(length);
        byte[] result = Arrays.copyOfRange(data, offset, offset + length);
        offset += length;
        return result;
    }

    // ==========================================================================
    // OPTION
    // ==========================================================================

    /**
     * Deserialize an optional value.
     *
     * @param deserializer function to deserialize the inner value
     * @param <T> the type of the optional value
     * @return the value or null if None
     */
    public <T> T readOption(Function<BcsDeserializer, T> deserializer) {
        ensureBytes(1);
        int tag = data[offset++] & 0xFF;
        if (tag == 0) {
            return null;
        }
        if (tag == 1) {
            return deserializer.apply(this);
        }
        throw BcsError.invalidOption(tag);
    }

    // ==========================================================================
    // VECTOR
    // ==========================================================================

    /**
     * Deserialize a vector of values.
     *
     * @param deserializer function to deserialize each element
     * @param <T> the type of vector elements
     * @return the list of values
     */
    public <T> List<T> readVector(Function<BcsDeserializer, T> deserializer) {
        long length = readUleb128();
        // Check both MAX_SEQUENCE_LENGTH and Integer.MAX_VALUE for defense-in-depth
        if (length > BcsSerializer.MAX_SEQUENCE_LENGTH || length > Integer.MAX_VALUE) {
            throw BcsError.exceededMaxLength(length);
        }
        int len = (int) length;
        List<T> result = new ArrayList<>(len);
        for (int i = 0; i < len; i++) {
            result.add(deserializer.apply(this));
        }
        return result;
    }

    // ==========================================================================
    // ENUM
    // ==========================================================================

    /**
     * Enter an enum container and read its variant index (ULEB128).
     *
     * @return the variant index
     * @throws BcsError if the variant index exceeds Integer.MAX_VALUE
     */
    public int enterEnum() {
        enterContainer("enum");
        long index = readUleb128();
        if (index > Integer.MAX_VALUE) {
            throw BcsError.valueOutOfRange("variant index", index);
        }
        return (int) index;
    }

    /**
     * Leave the current enum container.
     */
    public void leaveEnum() {
        leaveContainer();
    }

    /**
     * Read an enum variant index (ULEB128).
     *
     * @return the variant index
     */
    public int readVariantIndex() {
        return enterEnum();
    }

    // ==========================================================================
    // MAP
    // ==========================================================================

    /**
     * Deserialize a map (verifying sorted keys).
     *
     * @param keyDeserializer function to deserialize keys
     * @param valueDeserializer function to deserialize values
     * @param <K> the key type
     * @param <V> the value type
     * @return the map
     */
    public <K, V> Map<K, V> readMap(
            Function<BcsDeserializer, K> keyDeserializer, Function<BcsDeserializer, V> valueDeserializer) {

        long length = readUleb128();
        if (length > BcsSerializer.MAX_SEQUENCE_LENGTH) {
            throw BcsError.exceededMaxLength(length);
        }
        // Defense-in-depth: explicit check before cast
        if (length > Integer.MAX_VALUE) {
            throw BcsError.exceededMaxLength(length);
        }
        int len = (int) length;

        // Pre-allocate with capacity hint for better performance
        Map<K, V> result = new LinkedHashMap<>(len);
        byte[] prevKeyBytes = null;

        for (int i = 0; i < len; i++) {
            // Record position before reading key
            int keyStart = offset;
            K key = keyDeserializer.apply(this);
            int keyEnd = offset;

            // Verify key order using view comparison (zero-copy)
            if (prevKeyBytes != null) {
                int cmp = compareBytesRange(prevKeyBytes, 0, prevKeyBytes.length, 
                                           data, keyStart, keyEnd - keyStart);
                if (cmp == 0) {
                    throw BcsError.nonCanonicalMap("duplicate key");
                }
                if (cmp > 0) {
                    throw BcsError.nonCanonicalMap("keys not sorted");
                }
            }
            // Only copy for storing as previous key
            prevKeyBytes = Arrays.copyOfRange(data, keyStart, keyEnd);

            V value = valueDeserializer.apply(this);
            result.put(key, value);
        }

        return result;
    }

    // ==========================================================================
    // UTILITY
    // ==========================================================================

    /**
     * Check that all input has been consumed.
     *
     * @throws BcsError if there is remaining input
     */
    public void checkEnd() {
        if (offset < data.length) {
            throw BcsError.remainingInput(data.length - offset);
        }
    }

    /**
     * Get the number of remaining bytes.
     *
     * @return the number of remaining bytes
     */
    public int remaining() {
        return data.length - offset;
    }

    /**
     * Get the current offset.
     *
     * @return the current position
     */
    public int position() {
        return offset;
    }

    // ==========================================================================
    // CONTAINER DEPTH
    // ==========================================================================

    /**
     * Enter a struct container for depth tracking.
     */
    public void enterStruct() {
        enterStruct("");
    }

    /**
     * Enter a struct container for depth tracking.
     *
     * @param name optional struct name for error messages
     */
    public void enterStruct(String name) {
        enterContainer(name);
    }

    /**
     * Leave the current struct container.
     */
    public void leaveStruct() {
        leaveContainer();
    }

    // ==========================================================================
    // PRIVATE HELPERS
    // ==========================================================================

    private void ensureBytes(int count) {
        if (offset + count > data.length) {
            throw BcsError.unexpectedEof(count, data.length - offset);
        }
    }

    private void enterContainer(String container) {
        if (currentDepth >= BcsSerializer.MAX_CONTAINER_DEPTH) {
            throw BcsError.exceededContainerDepth(container);
        }
        currentDepth++;
    }

    private void leaveContainer() {
        if (currentDepth > 0) {
            currentDepth--;
        }
    }

    private BigInteger readBigIntegerLE(int byteLength) {
        // Convert little-endian bytes to BigInteger (big-endian, positive)
        // Allocate one extra byte at the start for sign (ensure positive)
        byte[] bigEndian = new byte[byteLength + 1];
        // bigEndian[0] = 0 by default, ensures positive BigInteger

        // Reverse bytes from little-endian to big-endian
        int srcEnd = offset + byteLength - 1;
        for (int i = 1; i <= byteLength; i++) {
            bigEndian[i] = data[srcEnd - i + 1];
        }
        offset += byteLength;
        return new BigInteger(bigEndian);
    }

    /**
     * Compare two byte arrays as unsigned bytes (lexicographic).
     */
    private static int compareBytes(byte[] a, byte[] b) {
        return Arrays.compareUnsigned(a, b);
    }

    /**
     * Compare a byte array with a range in another array (zero-copy comparison).
     */
    private static int compareBytesRange(byte[] a, int aOffset, int aLen, 
                                         byte[] b, int bOffset, int bLen) {
        int minLen = Math.min(aLen, bLen);
        for (int i = 0; i < minLen; i++) {
            int cmp = (a[aOffset + i] & 0xFF) - (b[bOffset + i] & 0xFF);
            if (cmp != 0) {
                return cmp;
            }
        }
        return aLen - bLen;
    }
}

File: sdks/csharp/Bcs/BcsDeserializer.cs
using System.Buffers.Binary;
using System.Numerics;
using System.Text;

namespace Bcs;

/// <summary>
/// BCS Deserializer - Manual deserialization API.
/// </summary>
/// <remarks>
/// Provides explicit methods for deserializing each BCS type.
/// Use this for full control over deserialization.
/// </remarks>
/// <example>
/// <code>
/// var des = new BcsDeserializer(bytes);
/// byte value1 = des.ReadU8();
/// ulong value2 = des.ReadU64();
/// string value3 = des.ReadString();
/// des.CheckEnd();
/// </code>
/// </example>
public class BcsDeserializer
{
    // Pre-computed constants for signed integer conversion
    private static readonly BigInteger SignBit128 = BigInteger.One << 127;
    private static readonly BigInteger TwoPow128 = BigInteger.One << 128;
    private static readonly BigInteger SignBit256 = BigInteger.One << 255;
    private static readonly BigInteger TwoPow256 = BigInteger.One << 256;

    private readonly ReadOnlyMemory<byte> _data;
    private int _offset;
    private int _depth;

    /// <summary>Creates a new BCS deserializer.</summary>
    public BcsDeserializer(ReadOnlyMemory<byte> data)
    {
        _data = data;
        _offset = 0;
        _depth = 0;
    }

    /// <summary>Creates a new BCS deserializer from a byte array.</summary>
    public BcsDeserializer(byte[] data) : this(data.AsMemory())
    {
    }

    #region Boolean

    /// <summary>Deserializes a boolean value.</summary>
    public bool ReadBool()
    {
        EnsureBytes(1);
        var b = _data.Span[_offset++];
        return b switch
        {
            0 => false,
            1 => true,
            _ => throw BcsException.InvalidBoolean(b)
        };
    }

    #endregion

    #region Unsigned Integers

    /// <summary>Deserializes an unsigned 8-bit integer.</summary>
    public byte ReadU8()
    {
        EnsureBytes(1);
        return _data.Span[_offset++];
    }

    /// <summary>Deserializes an unsigned 16-bit integer (little-endian).</summary>
    public ushort ReadU16()
    {
        EnsureBytes(2);
        var value = BinaryPrimitives.ReadUInt16LittleEndian(_data.Span.Slice(_offset));
        _offset += 2;
        return value;
    }

    /// <summary>Deserializes an unsigned 32-bit integer (little-endian).</summary>
    public uint ReadU32()
    {
        EnsureBytes(4);
        var value = BinaryPrimitives.ReadUInt32LittleEndian(_data.Span.Slice(_offset));
        _offset += 4;
        return value;
    }

    /// <summary>Deserializes an unsigned 64-bit integer (little-endian).</summary>
    public ulong ReadU64()
    {
        EnsureBytes(8);
        var value = BinaryPrimitives.ReadUInt64LittleEndian(_data.Span.Slice(_offset));
        _offset += 8;
        return value;
    }

    /// <summary>Deserializes an unsigned 128-bit integer (little-endian).</summary>
    public BigInteger ReadU128()
    {
        EnsureBytes(16);
        return ReadBigIntegerLE(16, false);
    }

    /// <summary>Deserializes an unsigned 256-bit integer (little-endian).</summary>
    public BigInteger ReadU256()
    {
        EnsureBytes(32);
        return ReadBigIntegerLE(32, false);
    }

    #endregion

    #region Signed Integers

    /// <summary>Deserializes a signed 8-bit integer (two's complement).</summary>
    public sbyte ReadI8()
    {
        return unchecked((sbyte)ReadU8());
    }

    /// <summary>Deserializes a signed 16-bit integer (two's complement, little-endian).</summary>
    public short ReadI16()
    {
        EnsureBytes(2);
        var value = BinaryPrimitives.ReadInt16LittleEndian(_data.Span.Slice(_offset));
        _offset += 2;
        return value;
    }

    /// <summary>Deserializes a signed 32-bit integer (two's complement, little-endian).</summary>
    public int ReadI32()
    {
        EnsureBytes(4);
        var value = BinaryPrimitives.ReadInt32LittleEndian(_data.Span.Slice(_offset));
        _offset += 4;
        return value;
    }

    /// <summary>Deserializes a signed 64-bit integer (two's complement, little-endian).</summary>
    public long ReadI64()
    {
        EnsureBytes(8);
        var value = BinaryPrimitives.ReadInt64LittleEndian(_data.Span.Slice(_offset));
        _offset += 8;
        return value;
    }

    /// <summary>Deserializes a signed 128-bit integer (two's complement, little-endian).</summary>
    public BigInteger ReadI128()
    {
        EnsureBytes(16);
        return ReadBigIntegerLE(16, true);
    }

    /// <summary>Deserializes a signed 256-bit integer (two's complement, little-endian).</summary>
    public BigInteger ReadI256()
    {
        EnsureBytes(32);
        return ReadBigIntegerLE(32, true);
    }

    #endregion

    #region ULEB128

    /// <summary>Deserializes a ULEB128-encoded unsigned integer.</summary>
    public uint ReadUleb128()
    {
        Uleb128.Decode(_data.Span.Slice(_offset), out var value, out var bytesConsumed);
        _offset += bytesConsumed;
        return value;
    }

    #endregion

    #region Bytes and Strings

    /// <summary>Deserializes a byte array (length-prefixed with ULEB128).</summary>
    public byte[] ReadBytes()
    {
        var length = ReadUleb128();
        if (length > BcsSerializer.MaxSequenceLength)
        {
            throw BcsException.ExceededMaxLength(length);
        }
        return ReadFixedBytes((int)length);
    }

    /// <summary>Deserializes a UTF-8 string (length-prefixed with ULEB128).</summary>
    public string ReadString()
    {
        var bytes = ReadBytes();
        try
        {
            return Encoding.UTF8.GetString(bytes);
        }
        catch (DecoderFallbackException)
        {
            throw BcsException.InvalidUtf8();
        }
    }

    /// <summary>Deserializes fixed-length bytes (no length prefix).</summary>
    public byte[] ReadFixedBytes(int length)
    {
        EnsureBytes(length);
        var result = _data.Span.Slice(_offset, length).ToArray();
        _offset += length;
        return result;
    }

    #endregion

    #region Option

    /// <summary>Reads the option tag and returns whether a value is present.</summary>
    public bool ReadOptionTag()
    {
        EnsureBytes(1);
        var tag = _data.Span[_offset++];
        return tag switch
        {
            0 => false,
            1 => true,
            _ => throw BcsException.InvalidOption(tag)
        };
    }

    /// <summary>Deserializes an optional value.</summary>
    public T? ReadOption<T>(Func<BcsDeserializer, T> deserializer) where T : class
    {
        return ReadOptionTag() ? deserializer(this) : null;
    }

    /// <summary>Deserializes an optional value type.</summary>
    public T? ReadOptionValue<T>(Func<BcsDeserializer, T> deserializer) where T : struct
    {
        return ReadOptionTag() ? deserializer(this) : null;
    }

    #endregion

    #region Vector

    /// <summary>Reads the length prefix for a vector.</summary>
    public uint ReadVectorLength()
    {
        var length = ReadUleb128();
        if (length > BcsSerializer.MaxSequenceLength)
        {
            throw BcsException.ExceededMaxLength(length);
        }
        return length;
    }

    /// <summary>Deserializes a vector of values.</summary>
    public List<T> ReadVector<T>(Func<BcsDeserializer, T> deserializer)
    {
        var length = ReadVectorLength();
        var result = new List<T>((int)length);
        for (var i = 0u; i < length; i++)
        {
            result.Add(deserializer(this));
        }
        return result;
    }

    #endregion

    #region Enum

    /// <summary>Reads an enum variant index (ULEB128).</summary>
    public uint ReadVariantIndex()
    {
        return ReadUleb128();
    }

    #endregion

    #region Map

    /// <summary>Reads the length prefix for a map.</summary>
    public uint ReadMapLength()
    {
        return ReadVectorLength();
    }

    /// <summary>Gets the current position for key comparison.</summary>
    public int Position => _offset;

    /// <summary>Gets a slice of the data from start to current position.</summary>
    public ReadOnlySpan<byte> SliceFrom(int start) => _data.Span.Slice(start, _offset - start);

    /// <summary>Deserializes a map with key validation (sorted keys, no duplicates).</summary>
    /// <typeparam name="K">The key type.</typeparam>
    /// <typeparam name="V">The value type.</typeparam>
    /// <param name="keyDeserializer">Function to deserialize keys.</param>
    /// <param name="valueDeserializer">Function to deserialize values.</param>
    /// <returns>A dictionary containing the deserialized key-value pairs.</returns>
    /// <exception cref="BcsException">Thrown if keys are not sorted or contain duplicates.</exception>
    public Dictionary<K, V> ReadMap<K, V>(
        Func<BcsDeserializer, K> keyDeserializer,
        Func<BcsDeserializer, V> valueDeserializer) where K : notnull
    {
        var length = ReadMapLength();
        var result = new Dictionary<K, V>((int)length);
        byte[]? prevKeyBytes = null;

        for (var i = 0u; i < length; i++)
        {
            var keyStart = _offset;
            var key = keyDeserializer(this);
            var keyEnd = _offset;
            var keyBytes = _data.Span.Slice(keyStart, keyEnd - keyStart).ToArray();

            if (prevKeyBytes != null)
            {
                var cmp = CompareBytes(prevKeyBytes, keyBytes);
                if (cmp == 0)
                {
                    throw BcsException.DuplicateMapKey();
                }
                if (cmp > 0)
                {
                    throw BcsException.NonCanonicalMap();
                }
            }
            prevKeyBytes = keyBytes;

            var value = valueDeserializer(this);
            result[key] = value;
        }

        return result;
    }

    /// <summary>Compare two byte arrays lexicographically (unsigned).</summary>
    private static int CompareBytes(byte[] a, byte[] b)
    {
        var minLen = Math.Min(a.Length, b.Length);
        for (var i = 0; i < minLen; i++)
        {
            if (a[i] < b[i]) return -1;
            if (a[i] > b[i]) return 1;
        }
        return a.Length.CompareTo(b.Length);
    }

    #endregion

    #region Container Depth

    /// <summary>Gets the current container nesting depth.</summary>
    public int ContainerDepth => _depth;

    /// <summary>Enters a struct container for depth tracking.</summary>
    /// <exception cref="BcsException">Thrown when container depth exceeds MaxContainerDepth (500).</exception>
    public BcsDeserializer EnterStruct()
    {
        EnterContainer("struct");
        return this;
    }

    /// <summary>Leaves a struct container.</summary>
    public BcsDeserializer LeaveStruct()
    {
        LeaveContainer();
        return this;
    }

    /// <summary>Enters an enum container for depth tracking and reads the variant index.</summary>
    /// <returns>The enum variant index.</returns>
    /// <exception cref="BcsException">Thrown when container depth exceeds MaxContainerDepth (500).</exception>
    public uint EnterEnum()
    {
        EnterContainer("enum");
        return ReadVariantIndex();
    }

    /// <summary>Leaves an enum container.</summary>
    public BcsDeserializer LeaveEnum()
    {
        LeaveContainer();
        return this;
    }

    private void EnterContainer(string containerType)
    {
        if (_depth >= BcsSerializer.MaxContainerDepth)
        {
            throw BcsException.ExceededContainerDepth(containerType);
        }
        _depth++;
    }

    private void LeaveContainer()
    {
        if (_depth > 0)
        {
            _depth--;
        }
    }

    #endregion

    #region Utility

    /// <summary>Verifies that all input has been consumed.</summary>
    public void CheckEnd()
    {
        if (_offset < _data.Length)
        {
            throw BcsException.RemainingInput(_data.Length - _offset);
        }
    }

    /// <summary>Gets the number of remaining bytes.</summary>
    public int Remaining => _data.Length - _offset;

    #endregion

    #region Private Helpers

    private void EnsureBytes(int count)
    {
        if (_offset + count > _data.Length)
        {
            throw BcsException.UnexpectedEof(count, _data.Length - _offset);
        }
    }

    private BigInteger ReadBigIntegerLE(int byteLength, bool signed)
    {
        // Read directly from span without allocation
        var span = _data.Span.Slice(_offset, byteLength);
        _offset += byteLength;

        var value = new BigInteger(span, isUnsigned: true, isBigEndian: false);

        if (signed)
        {
            // Use pre-computed constants for common sizes
            var (signBit, twoPow) = byteLength switch
            {
                16 => (SignBit128, TwoPow128),
                32 => (SignBit256, TwoPow256),
                _ => (BigInteger.One << (byteLength * 8 - 1), BigInteger.One << (byteLength * 8))
            };

            if (value >= signBit)
            {
                value -= twoPow;
            }
        }

        return value;
    }

    #endregion
}
